dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,199 @@ public Pattern getParameterPattern() {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
+        }
+    }
+
+    private class DatabaseProcedure extends DatabaseExecutable {
+
+        private DatabaseProcedure(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getProcedures(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseExecutable {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseObjectName {
+        String schemaName, objectName;
+
+        private DatabaseObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    DatabaseObjectName createDbObjectNameWithCurrentDefaultSchema(String objName) throws SQLException {
+        return new DatabaseObjectName(getConnection().getSchema(), objName);
```

Comment:
> Do you mean that data-like classes should just really have a constructor that assigns values to internal fields and getters to retrieve them?

Yes, if this is simpler, trivial and obvious representation which does the job.

Other points might be: 

* In sense of simplicity - avoid mixing multiple responsibilities.
* Consistency with other common API (e.g. [QName](https://docs.oracle.com/javase/8/docs/api/javax/xml/namespace/QName.html), [Integer](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html), [LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html) - even though these have other additional stronger reasons to that approach)


Indeed, it's fairly subjective matter and the parsing is simple. It's hard to make a strong case (e.g. like in http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/).

What matters now is the current context so if things look better as they are - no problem to be that way.